### ISSUE TITLE
Hostess & Waitress should match shiny::Progress methods

### ISSUE THREAD

**Issue by hadley** [JohnCoene/waiter#23]
As much as possible, making it easy to switch from the built-in progress bar to the waiter progress bar.

**Comment by JohnCoene** [comment]
I have added the `close` and `inc` method, identical to `hide` and `increase` respectively, and deprecated the latter the two. Method names now match `shiny::Progress` one-to-one.

**Comment by hadley** [comment]
Wonderful, thanks!

**Comment by hadley** [comment]
BTW in progress you can set `min` and `max` on creation so that `inc()` isn't necessarily a percentage or proportion.

**Comment by hadley** [comment]
Also note that you use `$set()` to start a `Progress` bar, but I've filed an issue on Shiny to encourage them to add a `start()` method: https://github.com/rstudio/shiny/issues/2724

(To be clear, I don't think you need to add a matching `message` argument since it's not part of the Waitress UI)

**Comment by hadley** [comment]
(Also, please feel free to request PRs from me; I'm very happy to contribute if it would be helpful)

**Comment by JohnCoene** [comment]
I've added `min` and `max` arguments which are passed when creating the waitress, I had not thought of that, great idea! 

However, I'm unsure about `start`. From the point of view of the user it would seem "logical" that `set` can be called multiple times but `start` only once. Or at least that would be my assumption.

It might make more sense for `shiny::Progress` as one "starts" the progress with a message to then incrementally "set" the value of the progress bar. 

Unless I am missing something, which could very well be.

```r
library(shiny)
library(waiter)

ui <- fluidPage(
  use_waitress(),
  sliderInput("value", "Waitress value", value = 50, min = 20, max = 500),
  div(id = "DIV")
)

server <- function(input, output){

  w <- Waitress$new("#DIV", min = 20, max = 500) 

  observeEvent(input$value, {
    w$set(input$value)
  })

}

shinyApp(ui, server)
```

**Comment by JohnCoene** [comment]
Re "feel free to request PRs from me." Thank you very much! I don't think the issues with the package currently warrant this, plus I'm learning a lot fixing them. I'm already very thankful to you for spending the time looking at the package, it's genuinely invaluable.

**Comment by JohnCoene** [comment]
The hostess now also matches `shiny::Progress` in argument names and methods. I spoke much too soon about the `start` method; it comes in handy.

Also if `waiter` has a `hide_on_render` argument, so should the waitress;  this has thus been added.

**Comment by JohnCoene** [comment]
I've added "features" to the hostess, the underlying Javascript library ([loading bar](https://loading.io/progress/)), gives a great deal of customisation of the loading bars. So far only the standard loading bar and the "presets" were available.

I've added a few arguments that allow doing "fancy" the likes of:

:warning: The snippet below creates an `assets` directory.

```r
library(shiny)
library(waiter)

dir.create("assets")
download.file("https://raw.githubusercontent.com/rstudio/hex-stickers/master/SVG/bookdown.svg", destfile = "./assets/bookdown.svg")
shiny::addResourcePath("assets", "assets")

ui <- fluidPage(
	use_waiter(),
	use_hostess(),
	actionButton("draw", "redraw"),
	plotOutput("plot")
)

server <- function(input, output) {

	dataset <- reactive({
		input$draw

		hostess <- Hostess$new(min = 0, max = 10)
		hostess$set_loader <- hostess_loader(svg = "assets/bookdown.svg", progress_type = "fill")
		waiter <- Waiter$new(
			"plot", 
			hostess$loader()
		)

		waiter$show()

		for(i in 1:10){
			Sys.sleep(.2)
			hostess$inc(1)
		}

		runif(100)

	})

	output$plot <- renderPlot(plot(dataset()))

}

shinyApp(ui, server)

# unlink("assets", recursive = TRUE)
``` 

![hostess-bookdown](https://user-images.githubusercontent.com/7833796/70662183-071ffc00-1c66-11ea-8678-f45176f64179.gif)

_the real thing is not choppy like the gif above_

or:

```r
library(shiny)
library(waiter)

ui <- fluidPage(
	use_waiter(),
	use_hostess(),
	actionButton("draw", "redraw"),
	plotOutput("plot")
)

server <- function(input, output) {

	dataset <- reactive({
		input$draw

		hostess <- Hostess$new(min = 0, max = 10)
		hostess$set_loader <- hostess_loader(
			progress_type = "stroke",
			stroke_color = hostess_stripe("red", "white")
		)
		waiter <- Waiter$new(
			"plot", 
			hostess$loader()
		)

		waiter$show()

		for(i in 1:10){
			Sys.sleep(.2)
			hostess$inc(1)
		}

		runif(100)

	})

	output$plot <- renderPlot(plot(dataset()))

}

shinyApp(ui, server)
```

![hostess-stripe](https://user-images.githubusercontent.com/7833796/70662520-a9d87a80-1c66-11ea-903c-56b7b1f00f75.gif)

_tiny on the gif, gives a striped loading bar_

or even:

```r
library(shiny)
library(waiter)

heart <- paste0("M81.495,13.923c-11.368-5.261-26.234-0.311-31.489,11.032C44.74,13.612,29.879,8.657,18.511,13.923  ",
  "C6.402,19.539,0.613,33.883,10.175,50.804c6.792,12.04,18.826,21.111,39.831,37.379c20.993-16.268,33.033-25.344,39.819-37.379  ",
  "C99.387,33.883,93.598,19.539,81.495,13.923z")

ui <- fluidPage(
	use_waiter(),
	use_hostess(),
	actionButton("draw", "redraw"),
	plotOutput("plot")
)

server <- function(input, output) {

	dataset <- reactive({
		input$draw

		hostess <- Hostess$new(min = 0, max = 10)
		hostess$set_loader <- hostess_loader(
			progress_type = "stroke",
      svg = heart,
      stroke_color = "red"
		)
		waiter <- Waiter$new(
			"plot", 
			hostess$loader()
		)

		waiter$show()

		for(i in 1:10){
			Sys.sleep(.2)
			hostess$inc(1)
		}

		runif(100)

	})

	output$plot <- renderPlot(plot(dataset()))

}

shinyApp(ui, server)
```

![hostess-heart](https://user-images.githubusercontent.com/7833796/70664453-b8c12c00-1c6a-11ea-9b15-bcfa302baa81.gif)


The issue is that I believe is the API is becoming convoluted yet again. My assumption/recommendation is probably to use a single spinner/loading bar across the application (for pure aesthetic reasons), given the added functions and arguments it'd be painful to repeat those arguments across all waitresses/waiters. 

I thus added an active R6 field called `set_loader` which allows one to set a loader created with `hostess_loader`.

```r
my_loader <- hostess_loader()
h1 <- hostess$new()$set_loader <- my_loader
h2 <- hostess$new()$set_loader <- my_loader
```

So questions come to mind.

1. Well, those effects are fancy and all but are they truly recommended, should they be included at all?
2. Should it then be more similar to [set_theme](https://ggplot2.tidyverse.org/reference/theme_get.html) from ggplot2? Or should both be available?

**Comment by hadley** [comment]
It's a little frivolous, but it's very fun, and it's nice to give people the ability to express their creativity in some small way. So I think keep it, but don't go overboard in advertising it.

BTW I think you should either do `hostess$new()$loader <- my_loader` or `hostess$new()$set_loader(my_loader)` (and I personally prefer the second).

**Comment by JohnCoene** [comment]
Yes! I realised that `set_` "sounded" more like an action/function than a field but couldn't use `hostess$new()$loader <- my_loader` as `loader` _was_ already taken as method to retrieve the loader.

Went with `set_loader` as public method and conversely renamed the method to retrieve the loader to `get_loader`.

```r
loader <- hostess_loader()

h1 <- Hostess$
  new()$
  set_loader(loader)$
  get_loader()
``` 

Seems to make a lot more sense already. Thank you!

**Comment by JohnCoene** [comment]
I'm gonna go ahead and close this as I think everything is fixed, methods and arguments match `shiny::Progress`, feel free to re-open if it is not the case.
